### ISSUE TITLE
DOI fetching fails in a few cases

### ISSUE THREAD

**Issue by perrette** [perrette/myref#1]
The current method to retrieve DOI consists in search for regular expressions over the first two pages, and to keep the first one that appear.

Accepted prefixes are (lower or upper case):

    'doi:', 'doi: ', 'doi ', 'dx\.doi\.org/', 'doi/'

DOI itself is searched as:

    r"10\.\d\d\d\d/.*?"

And is expected to finish with:

    r"[, \n]"

The method fails in a few cases:
- when DOI spreads over two lines (e.g. [here](http://onlinelibrary.wiley.com/doi/10.1002/2014GL062611))
- when other DOIs appear before the actual paper's DOI, for example [here](http://www.sciencedirect.com/science/article/pii/S0921818105000706)

These could be solved by more permissive parsing of DOI, but keep it conservative for now until a good solution is found.

Nevertheless, existing edits / fixes currently include:
- underscore sometimes gets converted into an empty space by pdftotxt, so we also detect ending with any space followed by a digit. This solves at least one case.

**Comment by boyanpenkov** [comment]
https://github.com/MicheleCotrufo/pdf2doi might be a candidate, just to continue the conversation from https://github.com/perrette/papers/issues/28

**Comment by perrette** [comment]
Yes. Thanks for the suggestion. If you end up exploring that sort of things (with your large PDF database to test with !) I'd be glad if you could report back about what works best. And who knows, perhaps someone shows up who feels like merging all the good tools into @MicheleCotrufo's pdf2doi (if practical) or another stand-alone package. That should be a library with python bindings, ideally not a verbose command line tool (so that it can be used in other command line tools), though one could certainly call it with subprocess (as is already done here with poppler utils).

**Comment by perrette** [comment]
I did try that tool and it takes an order of magnitude longer to extract the DOI in the few cases I tested. I also checked on a reported case of failure (https://github.com/MicheleCotrufo/pdf2doi/issues/23) and `papers extract` succeeds when pdf2doi fails... Perhaps I got lucky (or unlucky?) but it seems `papers extract` is doing a better job so far. Too bad we'll have to wait before outsourcing that job.
